src/components/TodoItem.tsx

```diff
@@ -0,0 +1,197 @@
+
+    const updatedTodoIndex = todos
+      .findIndex((editedtodo) => editedtodo.id === todoId);
+
+    if (editedTodo !== null) {
+      editTodo(todoId, updatedTodoWithNewTitle)
+        .then((res) => {
+          const updatedTodos = [...todos];
+
+          updatedTodos[updatedTodoIndex] = res;
+
+          setTodos(updatedTodos);
+        })
+        .catch((error) => {
+          handleError('Unable to update a todo');
+          // eslint-disable-next-line no-console
+          console.log(error);
+        })
+        .finally(() => {
+          setIsEditing(false);
+          setIsLoading(false);
+          titleInputRef.current?.focus();
+        });
+    }
+  };
+
+  return (
+    <div data-cy="Todo" className={`${todo?.completed ? 'todo completed' : 'todo'}`} key={todo?.id}>
+      <label className="todo__status-label">
+        <input
+          data-cy="TodoStatus"
+          type="checkbox"
+          className="todo__status"
+          checked={todo?.completed}
+          onChange={() => handleStatusChange(todo?.id)}
+        />
+      </label>
+
+      {isEditing ? (
+        <>
+          <form
+            onSubmit={(e) => handleSubmit(e)}
+          >
+            <input
+              data-cy="TodoTitleField"
+              type="text"
+              className="todo__title-field"
+              placeholder="Empty todo will be deleted"
+              value={newTitle}
+              // defaultValue={todo.title}
+              ref={editedRef}
+              onChange={(e) => setNewTitle(e.target.value)}
+              onBlur={() => {
+                handleSaveChanges(todo.id);
+              }}
+              onKeyUp={(e) => {
+                if (e.key === 'Enter') {
+                  e.preventDefault();
+                  handleSaveChanges(todo.id);
+                }
+
+                if (e.key === 'Escape') {
+                  e.preventDefault();
+                  setIsEditing(false);
+                  setNewTitle(todo.title);
+                }
+              }}
+            />
+          </form>
+        </>
+
+      ) : (
+        <>
+          <span
+            data-cy="TodoTitle"
+            className="todo__title"
+            onDoubleClick={() => {
+              setIsEditing(true);
+              setIsEditedTodo(todo);
+            }}
+          >
+            {todo?.title}
+          </span>
+          <button
+            type="button"
+            className="todo__remove"
+            data-cy="TodoDelete"
+            onClick={() => handleDelete(todo?.id)}
+          >
+            ×
+          </button>
+        </>
+      )}
+
+      <div
+        data-cy="TodoLoader"
+        className={cn('modal overlay',
+          {
+            'is-active': (isDeleting === true)
+          || (todo.id === 0) || (isToggled && todo.id === toggledId)
+          || (isToggledAll) || (isLoading)
+          || ((todo.completed === true) && isGroupDeleting),
```

Comment:
The whole logic here looks really complicated, It's hard to read and probably does not work as intended. Try to refactor this code. If the logic needs to be so complicated then extract it to function outside `return` and just call it here.